cacao-related/cacao/man/cacao.1
.TH CACAO 1 "November 23, 2005"
.UC
.SH NAME
cacao \- a Java just-in-time compiler
.SH SYNOPSIS
.B cacao 
[
.I options
]
.I classname
[
.I argument ... 
]
.br
.B cacao
[
.I options
]
.B -jar
.I file.jar
[
.I argument ...
]
.br
.SH DESCRIPTION
The method \fBmain\fP of the class \fIclassname\fP is executed.
Method \fBmain\fP must have the signature \fBpublic static void
main(String[] argv)\fP. The program arguments are passed to method
\fBmain\fP in the string array \fBargv\fP. The environment variable
\fICLASSPATH\fP must contain the directory of the Java class files.
.SH OPTIONS
Currently recognized options are:

.TP
.B -classpath \fIpath\fP
Set the search path for user-defined class files. Usually the environment variable CLASSPATH
specifies the search path, but this option overrides the value of the environment variable.
\fIpath\fP may be a colon-separated list of directories and .jar files.

.TP
.B -cp \fIpath\fP
Equivalent to \fB-classpath \fIpath\fP.

.TP
.B -D\fIproperty\fB=\fIvalue\fP
Assign a value to an entry of the system property list.

.TP
.BR -verbose [ :class | gc | jni ]
Enable specific verbose output.

.TP
.B -version
Print product version and exit.

.TP
.B -fullversion
Print jpackage-compatible product version and exit.

.TP
.B -showversion
Print product version and continue.

.TP
.B -help\fP, \fB-?
Print this help message and exit.

.TP
.B -X
Print help on non-standard Java options.

.SH NON-STANDARD OPTIONS
Currently recognized options are:

.TP
.B -Xbootclasspath/a:\fIpath\fP
Append \fIpath\fP to the search path of the bootstrap loader.

.TP
.B -Xbootclasspath/p:\fIpath\fP
Prepend \fIpath\fP to the search path of the bootstrap loader.

.TP
.B -Xbootclasspath:\fIpath\fP
Set the search path of the bootstrap loader to \fIpath\fP.
\fIpath\fP is a list of .zip files, .jar files, or directories, separated by \fB:\fP.

.TP
.B -Xint
Interpreter mode execution.

.TP
.B -Xjit
JIT mode execution (default).

.TP
.B -Xms\fIsize\fP
Set the initial size of the heap (default: 2MB).

.TP
.B -Xmx\fIsize\fP
Set the maximum size of the heap (default: 64MB).

.TP
.B -Xss\fIsize\fP
Set the thread stack size (default: 128kB).

.TP
.B -all
Compile all methods, no execution.

.TP
.B -cb
Disable array bounds checks.

.TP
.B -cs
Disable synchronization.

.TP
.B -l
Don't start the class after loading.

.TP
.B -liberalutf
Don't warn about overlong UTF-8 sequences.

.TP	
.B -log \fIlogfile\fP
Specifies the logfile. Default is \fIstdout\fP.

.TP
.B -lsra
Use linear scan register allocation.

.TP
.B -m \fImethod\fP
Compile only a specific method.

.TP
.B -noverify
Turn off bytecode verification and other verifier checks.

.TP
.B -sig \fIsignature\fP
Specify signature for the method given by the \fB-m\fP option.

.TP
.B -softnull
Use software nullpointer check.

.TP
.B -stat
Print detailed compiler statistics.

.TP	
.B -time
Gives statistics about run time after the program has completed.

.TP
.B -v
Print state information.

.SH DEBUGGING OPTIONS

The following options are useful for debugging \fBcacao\fP:

.TP
.B -sa
Show a disassembled listing of the generated code.

.TP
.B -sc
Show the constant pool.

.TP
.B -sd 
Show data segment listing.

.TP
.B -se
Show disassembled exception stubs (only with \fB-sa\fP).

.TP
.B -si
Show intermediate representation.

.TP
.B -sm
Show class fields and methods.

.TP
.B -sn
Show disassembled native stubs.

.TP
.B -su
Show the internal utf hash.

.TP
.B -verbose
Print more information.

.TP
.B -verbosecall
Trace method calls.

.TP
.B -verboseexception
Trace exceptions and stack unwinding.

.TP
.B -verbosegc
Print a message for each garbage collection.

.SH EXPERIMENTAL OPTIONS

.TP
.B -eager
Perform eager class loading and linking.

.TP
.B -ie
Inline methods with exceptions.

.TP
.B -in
Activate inlining of methods.

.TP
.B -io
Inline methods of foreign classes.

.TP
.B -ip
Optimize argument renaming when inlining.

.TP
.B -iv
Inline virtual methods (uses/turns \fB-rt\fP option on).

.TP
.B -oloop
Optimize array accesses in loops.

.TP
.B -rt
Use rapid type analysis.

.TP
.B -vta
Use variable type analysis.

.TP
.B -xta
Use X type analysis.

.SH ENVIRONMENT VARIABLES

.TP
.B CLASSPATH
A colon-separated list of directories and .jar files to search for
user-defined classes.

.TP
.B BOOTCLASSPATH
A colon-separated list of directories and .jar files searched
by the bootstrap loader.

.SH AUTHOR

The CACAO Team:

Reinhard Grafl, 
Andreas Krall, 
Christopher Kruegel, 
Carolyn Oates, 
Roman Obermaisser, 
Martin Platter, 
Mark Probst, 
Stefan Ring, 
Edwin Steiner, 
Christian Thalinger, 
Dieter Thuernbeck, 
Philipp Tomsich, 
Christian Ullrich, 
Joseph Wenninger 

Send mail to cacao@cacaojvm.org



ChangeLog
2007-10-14

    Start Onnama-2007

2007-09-24

    Don't kill the sim on div-by-zero and other changes

2007-09-22

    swsim: new option --disass-usercode
    load multiple elf files
    Simulator: Factored out the architectural MIPS state
    Initialize memory now matching endianess
    Speed stats.

2007-09-22  Tommy Thorn  <tommy@silver.local>

	* swsim: more prepping for cp0 and cp1 support.

	* swsim: can load multiple elf files at once.

        * swsim: Factored out the architectural MIPS state. Preparing
          for more CP0 and CP1 support.

        * swsim: Rough speed statistics

          Numbers from one test run: ISA simulation is close to twice
          as fast as real time, which ISA/RTL cosimulation is about
          16,000 times slower. 4% of the instructions are JAL and it
          averaged 2.65 CPI.

2007-09-15

    Be more explicit about unhandled instructions. Don't double load
    from IO space

2007-09-14

    Status update

2007-09-11

    Handle little-endian. Disassemble regs as $31 instead of r31, etc
    Prune unused code

2007-09-10

    Fixed a bad bug in the I$ interface. Now the "showtime" graphical
    demo loads with tinymon and executes correctly. Bring out the
    champagne.


2007-09-09

    The IO event counter is only 8-bit
    Easy control over verbosity
    Fix think-o in the simulated IO
    No timeouts on RTL simulations
    runmips was loading too many sections
    Minor build cleanup
    Minor cleanup
    Handle stalls directly
    Updated (and tested) the cross tool build script
    Emit the last 32 RTL traces upon divergence

2007-09-07

    Make ISA and RTL simulations match on IO behaviour

2007-09-06

    First feable ISA/RTL co-simulation running
    Dump all sections of ELF files

2007-09-04

    Changes needed for Mac OS X
    Make swsim not depend on system elf.h
    Silly re-indentation

2007-09-03

    Stupid bug in tinymon
    Verbosity killed the transfer. Silence is gold
    Updated .mif file for the quick hack
    Quick hack to see what's going on with my serial transmissions
    More small examples
    Example program changes

    Runmips knows about firmware simulation but still supports
    reformatting MIPS ELF files

    Tinymon more verbose about errors
    ISA simulation simulate uninitialized SRAM
    Firmware must initialize all data itself
    Better serial IO simulation in RTL
    _gp bugfix and Don't Drop Serial Input
    Much improved simulator tracing
    Stuff destined for read-only firmware needs to allocate data in
    SRAM

2007-09-02

    Cleaner clean
    Tinymon support readback
    Cleaner clean
    Deal more gracefully with DOS line conventions

2007-09-01

    Added emulated serial input to the RTL simulation and fixed a
    silly bug in tinymon.c

    Fixed a tricky bug in the cache line fetch

    Clearly my interconnect structure is too tricky for even me. I
    managed to use it incorrectly (Repeat after me: data is valid the
    cycle after strobe is set and wait is clear). This will all be
    replaced eventually.

    Fix indentation
    Add self-checking to Tinymon
    Debug message adjustment
    Null workload
    Remove a 4096 fudge factor - rationale lost in time
    Buglet fixed

2007-08-26

    Life is difficult without a writable text segment
    Broken checksums
    Make tinymon more tolerant
    ugh, are we dropping serial bytes?
    typo fix
    Even more feedback from tinymon
    More feedback from tinymon
    Add precompiled test case
    Promote tinymon to be the new default firmware.
    Fixed few minor issues.

    A functional tinymon and support in runmips to generate the
    format. Changes to runmips to support serial input from files.

    Update the serial interface (now 57600 bps)

2007-08-25

    Small fixes to the simulator, small fixes to testall.S

2007-08-23

    Fix a simulation problem and move toward a synchronous register file

    AFAICT, Icarus Verilog mistreats non-blocking assignments when mixed
    with blocking assignments, but the logic worked in Quartus II 7.1.
    Never the less, the rewritten Verilog is clearer and being able to
    test with Icarus is important.

    Minor changes:
    - Use the numeric registers when dissassembling.
    - A vain attempt to move towards a synchronous register file.
      Alas, it will take a significant rewrite to fix this. Punting on
      that as the whole pipeline will be replace "soon".

2007-08-22

    Fixed a compilation issue. *THIS* one sent to Martin

2007-08-22

    sent to Martin Schoeberl

2007-08-22

    Removed more junk

2007-08-21

    Cleaned up a bit

2007-08-21

    Onnama-2006 as it was left
    Development has resumed using GIT as the SCM.

0.19 2005-06-24
        - many moons later, started keeping Onnama under Hg which
          forces me to be more disciplined about what I write.
        - Check the README for the current state of affairs.

0.2 2004-01-30
        - signed/unsigned byte/short loads and stores implemented
        - first semi-public release

0.1 2004-01-28
        - Shifts were broken, but now Sieve works with bit strings!

0.0
        - Yeah, Sieve runs, spitting out primes on the serial port!



cacao-related/cacao/src/cacao/cacao.c
/* src/cacao/cacao.c - contains main() of cacao

   Copyright (C) 1996-2005, 2006, 2007 R. Grafl, A. Krall, C. Kruegel,
   C. Oates, R. Obermaisser, M. Platter, M. Probst, S. Ring,
   E. Steiner, C. Thalinger, D. Thuernbeck, P. Tomsich, C. Ullrich,
   J. Wenninger, J. Wenninger, Institut f. Computersprachen - TU Wien

   This file is part of CACAO.

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2, or (at
   your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301, USA.

   $Id: cacao.c 7698M 2007-12-20 15:22:17Z (local) $

*/


#include "config.h"

#include <assert.h>

#if defined(ENABLE_LIBJVM)
# include <ltdl.h>
#endif

#if defined(WITH_JRE_LAYOUT)
# include <errno.h>
# include <libgen.h>
# include <unistd.h>
#endif

#include <stdlib.h>
#include <string.h>

#include "vm/types.h"

#include "native/jni.h"

#if defined(ENABLE_JVMTI)
# include "native/jvmti/jvmti.h"
# include "native/jvmti/cacaodbg.h"
#endif

#include "vm/vm.h"


/* forward declarations *******************************************************/

static JavaVMInitArgs *cacao_options_prepare(int argc, char **argv);


/* main ************************************************************************

   The main program.
   
*******************************************************************************/

int main(int argc, char **argv)
{
#if defined(ENABLE_LIBJVM) && !defined(WITH_STATIC_CLASSPATH)
	char           *path;
#endif

#if defined(ENABLE_LIBJVM)	
	/* Variables for JNI_CreateJavaVM dlopen call. */
	lt_dlhandle     libjvm_handle;
	lt_ptr          libjvm_vm_createjvm;
	lt_ptr          libjvm_vm_run;
	const char     *lterror;

	bool (*vm_createjvm)(JavaVM **, void **, void *);
	void (*vm_run)(JavaVM *, JavaVMInitArgs *);
#endif

	JavaVM         *vm;                 /* denotes a Java VM                  */
	JNIEnv         *env;
	JavaVMInitArgs *vm_args;

	/* prepare the options */

	/* vm_args = cacao_options_prepare(argc, argv); */
/*	char *tmp[] = {"cacao", "hello"};
	vm_args = cacao_options_prepare(2, &tmp); */
	char *tmp[] = {"cacao", "jbe.DoKernel"};
	vm_args = cacao_options_prepare(2, &tmp); 
	
	
	/* load and initialize a Java VM, return a JNI interface pointer in env */

#if defined(ENABLE_LIBJVM) && !defined(WITH_STATIC_CLASSPATH)
# if defined(WITH_JRE_LAYOUT)
	/* SUN also uses a buffer of 4096-bytes (strace is your friend). */

	path = malloc(sizeof(char) * 4096);

	if (readlink("/proc/self/exe", path, 4095) == -1) {
		fprintf(stderr, "main: readlink failed: %s\n", strerror(errno));
		abort();
	}

	/* get the path of the current executable */

	path = dirname(path);

	if ((strlen(path) + strlen("/../lib/libjvm") + strlen("0")) > 4096) {
		fprintf(stderr, "main: libjvm name to long for buffer\n");
		abort();
	}

	/* concatinate the library name */

	strcat(path, "/../lib/libjvm");
# else
	path = CACAO_LIBDIR"/libjvm";
# endif

	if (lt_dlinit()) {
		fprintf(stderr, "main: lt_dlinit failed: %s\n", lt_dlerror());
		abort();
	}

	/* First try to open where dlopen searches, e.g. LD_LIBRARY_PATH.
	   If not found, try the absolute path. */

	if (!(libjvm_handle = lt_dlopenext("libjvm"))) {
		/* save the error message */

		lterror = strdup(lt_dlerror());

		if (!(libjvm_handle = lt_dlopenext(path))) {
			/* print the first error message too */

			fprintf(stderr, "main: lt_dlopenext failed: %s\n", lterror);

			/* and now the current one */

			fprintf(stderr, "main: lt_dlopenext failed: %s\n", lt_dlerror());
			abort();
		}

		/* free the error string */

		free((void *) lterror);
	}

	if (!(libjvm_vm_createjvm = lt_dlsym(libjvm_handle, "vm_createjvm"))) {
		fprintf(stderr, "main: lt_dlsym failed: %s\n", lt_dlerror());
		abort();
	}

	vm_createjvm =
		(bool (*)(JavaVM **, void **, void *)) (ptrint) libjvm_vm_createjvm;
#endif

	/* create the Java VM */

	(void) vm_createjvm(&vm, (void *) &env, vm_args);

#if defined(ENABLE_JVMTI)
	pthread_mutex_init(&dbgcomlock,NULL);
	if (jvmti) jvmti_set_phase(JVMTI_PHASE_START);
#endif

#if !defined(WITH_STATIC_CLASSPATH) && defined(ENABLE_LIBJVM)
	if (!(libjvm_vm_run = lt_dlsym(libjvm_handle, "vm_run"))) {
		fprintf(stderr, "lt_dlsym failed: %s\n", lt_dlerror());
		abort();
	}

	vm_run = (void (*)(JavaVM *, JavaVMInitArgs *)) (ptrint) libjvm_vm_run;
#endif

	/* run the VM */

	vm_run(vm, vm_args);

	/* keep compiler happy */

	return 0;
}


/* cacao_options_prepare *******************************************************

   Prepare the JavaVMInitArgs.

*******************************************************************************/

static JavaVMInitArgs *cacao_options_prepare(int argc, char **argv)
{
	JavaVMInitArgs *vm_args;
	s4              i;

	vm_args = malloc(sizeof(JavaVMInitArgs));

	vm_args->version            = JNI_VERSION_1_2;
	vm_args->nOptions           = argc - 1;
	vm_args->options            = malloc(sizeof(JavaVMOption) * argc);
	vm_args->ignoreUnrecognized = JNI_FALSE;

	for (i = 1; i < argc; i++)
		vm_args->options[i - 1].optionString = argv[i];

	return vm_args;
}


/*
 * These are local overrides for various environment variables in Emacs.
 * Please do not remove this and leave it at the end of the file, where
 * Emacs will automagically detect them.
 * ---------------------------------------------------------------------
 * Local variables:
 * mode: c
 * indent-tabs-mode: t
 * c-basic-offset: 4
 * tab-width: 4
 * End:
 */
